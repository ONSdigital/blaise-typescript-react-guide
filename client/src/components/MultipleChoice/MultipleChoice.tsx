import React, { useEffect, useState } from 'react';
import Checkbox from '../Checkbox';

interface MultipleChoiceProps {
  question: string
  values: string[]
  onUpdate: (items: string[]) => void
}

export default function MultipleChoice({ question, values, onUpdate }: MultipleChoiceProps) {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [selected, setSelected] = useState<string[]>([]);

  useEffect(() => {
    onUpdate(selected);
  }, [onUpdate, selected]);

  function selectValue(value: string) {
    setSelected((current) => [...current, value]);
  }

  function unselectValue(value: string) {
    setSelected((current) => current.filter((item) => item !== value));
  }

  return (
    <div className="multiple-choice">
      <div className="multiple-choice__question">{question}</div>
      <div className="multiple-choice__options">
        {values.map((value) => (
          <Checkbox
            key={value}
            label={value}
            onChecked={() => selectValue(value)}
            onUnchecked={() => unselectValue(value)}
          />
        ))}
      </div>
    </div>
  );
}
