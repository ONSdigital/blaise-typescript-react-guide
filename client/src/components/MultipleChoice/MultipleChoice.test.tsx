import React, {
  act, fireEvent, render, RenderResult, screen,
} from '@testing-library/react';
import MultipleChoice from './MultipleChoice';

describe('MultipleChoice', () => {
  let wrapper: RenderResult;
  let onUpdate: jest.Mock;

  beforeEach(() => {
    onUpdate = jest.fn();

    wrapper = render(
      <MultipleChoice
        question="question text"
        values={['Option 1', 'Option 2', 'Option 3']}
        onUpdate={onUpdate}
      />,
    );
  });

  it('displays the question', () => {
    expect(screen.getByText('question text')).toBeInTheDocument();
  });

  it('displays the options', () => {
    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Option 2')).toBeInTheDocument();
    expect(screen.getByText('Option 3')).toBeInTheDocument();
  });

  it('calls onUpdate when boxes are checked/unchecked', () => {
    act(() => {
      fireEvent.click(screen.getByLabelText('Option 1'));
      fireEvent.click(screen.getByLabelText('Option 3'));
      fireEvent.click(screen.getByLabelText('Option 1'));
    });
    expect(onUpdate).toHaveBeenCalledTimes(4);
    expect(onUpdate.mock.calls[0]).toContainEqual([]);
    expect(onUpdate.mock.calls[1]).toContainEqual(['Option 1']);
    expect(onUpdate.mock.calls[2]).toContainEqual(['Option 1', 'Option 3']);
    expect(onUpdate.mock.calls[3]).toContainEqual(['Option 3']);
  });

  test('snapshot', () => {
    act(() => {
      fireEvent.click(screen.getByLabelText('Option 1'));
    });
    expect(wrapper).toMatchSnapshot();
  });
});
