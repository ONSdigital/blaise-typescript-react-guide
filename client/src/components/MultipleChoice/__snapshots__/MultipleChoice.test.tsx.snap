// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleChoice snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="multiple-choice"
      >
        <div
          class="multiple-choice__question"
        >
          question text
        </div>
        <div
          class="multiple-choice__options"
        >
          <div
            class="checkbox"
          >
            <input
              class="checkbox__input"
              id=":r9:"
              type="checkbox"
            />
            <label
              class="checkbox__label"
              for=":r9:"
            >
              Option 1
            </label>
          </div>
          <div
            class="checkbox"
          >
            <input
              class="checkbox__input"
              id=":ra:"
              type="checkbox"
            />
            <label
              class="checkbox__label"
              for=":ra:"
            >
              Option 2
            </label>
          </div>
          <div
            class="checkbox"
          >
            <input
              class="checkbox__input"
              id=":rb:"
              type="checkbox"
            />
            <label
              class="checkbox__label"
              for=":rb:"
            >
              Option 3
            </label>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="multiple-choice"
    >
      <div
        class="multiple-choice__question"
      >
        question text
      </div>
      <div
        class="multiple-choice__options"
      >
        <div
          class="checkbox"
        >
          <input
            class="checkbox__input"
            id=":r9:"
            type="checkbox"
          />
          <label
            class="checkbox__label"
            for=":r9:"
          >
            Option 1
          </label>
        </div>
        <div
          class="checkbox"
        >
          <input
            class="checkbox__input"
            id=":ra:"
            type="checkbox"
          />
          <label
            class="checkbox__label"
            for=":ra:"
          >
            Option 2
          </label>
        </div>
        <div
          class="checkbox"
        >
          <input
            class="checkbox__input"
            id=":rb:"
            type="checkbox"
          />
          <label
            class="checkbox__label"
            for=":rb:"
          >
            Option 3
          </label>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
